crates/components/src/map_view.rs

```diff
@@ -473,39 +438,46 @@ impl MapView {
+                                            &update_state.graphics,
+                                            glam::vec2(event_size.x, event_size.y),
+                                        );
+                                        luminol_graphics::Event::new_standalone(
+                                            &update_state.graphics,
+                                            update_state.filesystem,
+                                            &viewport,
+                                            event,
+                                            &self.map.atlas,
+                                        )
+                                        .unwrap()
+                                        .unwrap()
+                                    });
+
                                 if response.rect.is_positive() {
                                     let clipped_rect =
                                         ui.ctx().screen_rect().intersect(response.rect);
-                                    let proj_rect = egui::Rect::from_min_size(
-                                        (ui.ctx().screen_rect().min - response.rect.min)
-                                            .max(Default::default())
-                                            .to_pos2(),
-                                        preview_sprite.sprite_size * clipped_rect.size()
-                                            / response.rect.size(),
+                                    let painter = luminol_graphics::Painter::new(
+                                        preview_sprite.prepare(&update_state.graphics),
                                     );
-                                    preview_sprite.set_proj(
-                                        &graphics_state.render_state,
-                                        glam::Mat4::orthographic_rh(
-                                            proj_rect.left(),
-                                            proj_rect.right(),
-                                            proj_rect.bottom(),
-                                            proj_rect.top(),
-                                            -1.,
-                                            1.,
+                                    ui.painter().add(
+                                        luminol_egui_wgpu::Callback::new_paint_callback(
+                                            clipped_rect,
+                                            painter,
                                         ),
                                     );
-                                    preview_sprite.paint(
-                                        graphics_state.clone(),
-                                        &painter,
-                                        clipped_rect,
-                                    );
                                 }
+
+                                self.preview_events.insert(event.id, preview_sprite);
                             }
```

Comment:
The new code for drawing event graphics in the hover UI needs to take into account the fact that `clipped_rect` might not represent the entire rectangle where the event graphic is supposed to appear. If the application window is too small for the entire event graphic to fit, `clipped_rect` will smaller than the actual size of the event graphic in order to get around wgpu's viewport size validation. That means in this case, you need to adjust for this by making the event graphic proportionally larger so that the perceived size of the event graphic remains the same despite the smaller viewport. That's what the purpose of `proj_rect` was in the original code. Without it, if the application window is smaller than the event graphic, the event graphic will appear squashed.

Also, I think `preview_events` and `last_events` can be removed. I don't see them being used anywhere other than here.

Reply:
> Also, I think preview_events and last_events can be removed. I don't see them being used anywhere other than here.

They were meant for a ping-pong system where preview sprites not used in the last frame were discarded '^^ 
I forgot to actually swap them in the code